src/dungeon_generation.c

```diff
@@ -0,0 +1,6448 @@
+                        break;
+                    case LAYOUT_OUTER_ROOMS:
+                        GenerateOuterRoomsFloor(gridSizeX, gridSizeY, floorProps);
+                        secondaryGen = TRUE;
+                        break;
+                }
+            }
+
+            ResetInnerBoundaryTileRows();
+            EnsureImpassableTilesAreWalls();
+
+            // Nothing failed during generation. This variable is always set to FALSE, so the check always passes
+            if (!sInvalidGeneration) {
+                // We need to make sure there are at least 2 rooms with at least 20 total tiles
+                s32 numRooms = 0;
+                bool8 rooms[64];
+                s32 roomTiles = 0;
+                s32 i;
+
+                for (i = 0; i < 64; i++) {
+                    rooms[i] = FALSE;
+                }
+
+                for (x = 0; x < DUNGEON_MAX_SIZE_X; x++) {
+                    for (y = 0; y < DUNGEON_MAX_SIZE_Y; y++) {
+                        const Tile *tile = GetTile(x, y);
+                        if ((tile->terrainType & (TERRAIN_TYPE_NORMAL | TERRAIN_TYPE_SECONDARY)) == TERRAIN_TYPE_NORMAL && tile->room <= 240) {
+                            roomTiles++;
+                            if (tile->room < 64) {
+                                rooms[tile->room] = TRUE;
+                            }
+                        }
+                    }
+                }
+
+                numRooms = 0;
+                // In order to match, 'y' had to be re-used as an iterator var
+                for (y = 0; y < 64; y++) {
+                    if (rooms[y]) {
+                        numRooms++;
+                    }
+                }
+
+                if (roomTiles >= 30 && numRooms >= 2) {
+                    break; // This layout is good!
+                }
+            }
+        }
+
+        // If we fail to generate a layout in 10 attempts, just abort and make a one-room Monster House
+        if (genAttempts == 10) {
+            sKecleonShopMiddlePos.x = -1;
+            sKecleonShopMiddlePos.y = -1;
+            GenerateOneRoomMonsterHouseFloor();
+            gDungeon->forceMonsterHouse = TRUE;
+        }
+
+        // We will be guaranteed to have a good layout by this point
+        FinalizeJunctions();
+        if (secondaryGen) {
+            GenerateSecondaryTerrainFormations(ROOM_FLAG_ALLOW_SECONDARY_TERRAIN, floorProps);
+        }
+
+        isEmptyMonsterHouse = (DungeonRandInt(100) < floorProps->itemlessMonsterHouseChance);
+        SpawnNonEnemies(floorProps, isEmptyMonsterHouse);
+        SpawnEnemies(floorProps, isEmptyMonsterHouse);
+
+        ResolveInvalidSpawns(); // Make sure multiple flags aren't set for one tile
+        if (gDungeon->playerSpawn.x != -1 && gDungeon->playerSpawn.y != -1) {
+            // This is for normal fixed rooms, we don't need to validate the stairs in this scenario
+			// Since it's fixed already
```

Comment:
Looks like tabs/spaces are mixed in this file.

Reply:
hopefully fixed now